fullstory-segment-middleware/src/main/java/com/fullstorydev/fullstory_segment_middleware/FullStorySegmentMiddleware.java

```diff
@@ -159,4 +167,118 @@ private BasePayload getNewPayloadWithFSURL(BasePayload payload, Map<String, Obje
         }
         return null;
     }
+
+    private Map<String, Object> getSuffixedProps (Map<?, ?> props) {
+        // transform props to comply with FS custom events requirement
+        // TODO: Segment should fail with circular dependency, but we should check anyway
+        Map<String, Object> mutableProps = new HashMap<>();
+        Stack<Map<?,?>> stack = new Stack<>();
+
+        stack.push(props);
+
+        while (!stack.empty()) {
+            Map<?,?> map = stack.pop();
+            for (Map.Entry<?, ?> entry : map.entrySet()) {
+                // We should only be getting String type keys, but parse the map to make sure the keys we use are all String
+                String key = String.valueOf(entry.getKey());
+                Object item = entry.getValue();
+
+                if (item instanceof Map) {
+                    // nested maps, concatenate keys and push back to stack
+                    Map<?,?> itemMap = (Map<?,?>) item;
+                    for (Map.Entry<?, ?> e : itemMap.entrySet()) {
+                        String concatenatedKey =  key + '.' + e.getKey();
+                        Map<String,Object> m = new HashMap<>();
+                        m.put(concatenatedKey, e.getValue());
+                        stack.push(m);
+                    }
+                } else if (item != null && item.getClass().isArray()) {
+                    // if it's of array type (but not iterable)
+                    // To comply with FS requirements, flatten the array of objects into a map:
+                    // each item in array becomes a map, with key, and item
+                    // enable search value the array in FS (i.e. searching for one product when array of products are sent)
+                    // then push each item with the same key back to stack
+                    Object[] objs = this.getArrayObjectFromArray(item);
+                    for (Object obj: objs) {
+                        Map<String,Object> m = new HashMap<>();
+                        m.put(key, obj);
+                        stack.push(m);
+                    }
+                } else if (item instanceof Iterable) {
```

Comment:
nit: slight readability improvement if you swap this test with the `else if (item != null && item.getClass().isArray()) {` test since the final `else` statement is capturing primitives and null items